clickhouse-admin/src/clickhouse_cli.rs

```diff
@@ -0,0 +1,130 @@
+
+#[derive(Debug, thiserror::Error, SlogInlineError)]
+pub enum ClickhouseCliError {
+    #[error("failed to run `clickhouse {subcommand}`")]
+    Run {
+        description: &'static str,
+        subcommand: String,
+        #[source]
+        err: io::Error,
+    },
+    #[error(transparent)]
+    ExecutionError(#[from] ExecutionError),
+    #[error("failed to parse command output")]
+    Parse {
+        description: &'static str,
+        stdout: String,
+        stderr: String,
+        #[source]
+        err: anyhow::Error,
+    },
+}
+
+impl From<ClickhouseCliError> for HttpError {
+    fn from(err: ClickhouseCliError) -> Self {
+        match err {
+            ClickhouseCliError::Run { .. }
+            | ClickhouseCliError::Parse { .. }
+            | ClickhouseCliError::ExecutionError(_) => {
+                let message = InlineErrorChain::new(&err).to_string();
+                HttpError {
+                    status_code: http::StatusCode::INTERNAL_SERVER_ERROR,
+                    error_code: Some(String::from("Internal")),
+                    external_message: message.clone(),
+                    internal_message: message,
+                }
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct ClickhouseCli {
+    /// Path to where the clickhouse binary is located
+    pub binary_path: Utf8PathBuf,
+    /// Address on where the clickhouse keeper is listening on
+    pub listen_address: SocketAddrV6,
+    pub log: Option<Logger>,
+}
+
+impl ClickhouseCli {
+    pub fn new(binary_path: Utf8PathBuf, listen_address: SocketAddrV6) -> Self {
+        Self { binary_path, listen_address, log: None }
+    }
+
+    pub fn with_log(mut self, log: Logger) -> Self {
+        self.log = Some(log);
+        self
+    }
+
+    pub async fn lgif(&self) -> Result<Lgif, ClickhouseCliError> {
+        self.keeper_client_non_interactive(
+            "lgif",
+            "Retrieve logically grouped information file",
+            Lgif::parse,
+            self.log.clone().unwrap(),
+        )
+        .await
+    }
+
+    async fn keeper_client_non_interactive<'a, F, T>(
+        &self,
+        query: &str,
+        subcommand_description: &'static str,
+        parse: F,
+        log: Logger,
+    ) -> Result<T, ClickhouseCliError>
+    where
+        F: FnOnce(&Logger, &[u8]) -> Result<T>,
+    {
+        let mut command = Command::new(&self.binary_path);
+        command
+            .arg("keeper-client")
+            .arg("--host")
+            .arg(&format!("[{}]", self.listen_address.ip()))
+            .arg("--port")
+            .arg(&format!("{}", self.listen_address.port()))
+            .arg("--query")
+            .arg(query);
+
+        let output = command.output().await.map_err(|err| {
+            let args: Vec<&OsStr> = command.as_std().get_args().collect();
+            let args_parsed: Vec<String> = args
+                .iter()
+                .map(|&os_str| os_str.to_str().unwrap().to_owned())
```

Comment:
So, `to_str` doesn't really return an error, it's an [Option](https://doc.rust-lang.org/std/ffi/struct.OsString.html#method.to_str), is the intention to return an error on None?. The only purpose of `args_parsed` is to provide information for the `ClickhouseCliError::Run` error. That's why I thought `to_str_lossy` made sense, that way we can see in the error exactly what arguments are being passed. So, in a way, I was already returning an error?

Maybe I'm misunderstanding, but it feels a bit strange to return an error because of a malformed error? It seems simpler to collect all the arguments, however they were passed and return a single error with that information like it was in the beginning with `to_str_lossy`. WDYT?

Reply:
TBH, we don't even need these `args_parsed` at all. I just thought it'd be useful to have them as part of the error message when returning an error, so an user could see exactly what arguments had been passed when running the command. The more I think about it, the more I think it's best to have `to_str_lossy` to see exactly what's being passed even in somehow a non-unicode character made it's way in somehow.

Unless this information isn't useful as part of the error message? In that case I could remove them altogether 🤔

Reply:
Ok, looks like I totally misinterpreted what was happening here. Sorry about that @karencfv! I didn't realize this was just for the error message. You were right using `to_string_lossy` in the first place. It's useful information to have, even if there's some weird non-utf8 char that gets excluded, which is doubtful anyway. That's better than panicking on unwrap. Sorry for the hassle.

Reply:
😄 nw! I've changed the names of the variables, so it's clearer what those are for